core/src/test/java/com/linecorp/armeria/server/ServerChildChannelPipelineCustomizerTest.java

```diff
@@ -0,0 +1,114 @@
+
+import javax.net.ssl.SSLEngine;
+
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.testng.util.Strings;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.ResponseTimeoutException;
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.testing.junit5.server.SelfSignedCertificateExtension;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.traffic.ChannelTrafficShapingHandler;
+
+class ServerChildChannelPipelineCustomizerTest {
+
+    @RegisterExtension
+    @Order(0)
+    static final SelfSignedCertificateExtension ssc = new SelfSignedCertificateExtension();
+
+    @RegisterExtension
+    @Order(1)
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            final SslContext sslCtx = SslContextBuilder
+                    .forServer(ssc.privateKey(), ssc.certificate())
+                    .startTls(true)
+                    .build();
+
+            sb.http(0);
+            sb.tlsSelfSigned();
+
+            sb.childChannelPipelineCustomizer(
+                    pipeline -> {
+                        final SSLEngine engine = sslCtx.newEngine(pipeline.channel().alloc());
+                        engine.setUseClientMode(false);
+                        pipeline.addFirst(new SslHandler(engine));
+                    });
+            sb.childChannelPipelineCustomizer(
+                    pipeline -> pipeline.addFirst(new ChannelTrafficShapingHandler(1024, 0)));
+
+            sb.service("/SslHandler", (ctx, req) -> HttpResponse.of(200));
+            sb.service("/TrafficShapingHandler", new AbstractHttpService() {
+                @Override
+                protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+                    return HttpResponse.of(
+                            ResponseHeaders.of(HttpStatus.OK, "header1", Strings.repeat("a", 2048)));
+                }
+            });
+        }
+    };
+
+    @Test
+    void testSessionProtocolNegotiationException() {
+        final ClientFactory clientFactory = ClientFactory.insecure();
+
+        final AggregatedHttpResponse response = WebClient
+                .builder(SessionProtocol.H1, server.httpEndpoint())
```

Comment:
Question: Does this work with H2 as well?

Reply:
This didn't work with H2.
The error message like the following was output.
```
com.linecorp.armeria.client.SessionProtocolNegotiationException: expected: h2, actual: h1, reason: unexpected protocol negotiation result
com.linecorp.armeria.client.UnprocessedRequestException: com.linecorp.armeria.client.SessionProtocolNegotiationException: expected: h2, actual: h1, reason: unexpected protocol negotiation result
```

Reply:
I got it. I guess it's because we didn't add sni handler. 😄